google/cloud/storage/internal/grpc/make_cord_test.cc

```diff
@@ -0,0 +1,82 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/storage/internal/async/write_payload_impl.h"
+#include "google/cloud/internal/absl_str_join_quiet.h"
+#include "google/cloud/internal/random.h"
+#include <gmock/gmock.h>
+
+namespace google {
+namespace cloud {
+namespace storage_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+
+using ::testing::ElementsAre;
+
+TEST(MakeCord, IsPayloadType) {
+  EXPECT_TRUE(IsPayloadType<char>::value);
+  EXPECT_TRUE(IsPayloadType<std::uint8_t>::value);
+  EXPECT_FALSE(IsPayloadType<std::string>::value);
+  EXPECT_FALSE(IsPayloadType<std::uint32_t>::value);
+}
+
+TEST(MakeCord, FromString) {
+  auto const input = std::string("The quick brown fox jumps over the lazy dog");
+  auto actual = MakeCord(input);
+  EXPECT_EQ(static_cast<std::string>(actual), input);
+}
+
+TEST(MakeCord, FromStringLong) {
+  auto const input = std::string(4 * 1024 * 1024, '\0');
+  auto actual = MakeCord(input);
+  std::vector<absl::string_view> chunks{actual.chunk_begin(),
+                                        actual.chunk_end()};
+  EXPECT_THAT(chunks, ElementsAre(absl::string_view(input)));
```

Comment:
string_view equality is based on value.  Do you want to check that the actual chunk really refers to the input?

Reply:
I think such a test would depend on `std::string(std::string&&)` actually moving the pointer, and typically we avoid such tests.

Reply:
I guess in this case it would even fail because `MakeCord(input)` makes a copy of `input`.

What I'm suggesting is a test for `MakeCord()`'s promise to _"create an `absl::Cord`, without copying the data in p"_.  That is, something like ...
```
  auto input = std::string(4 * 1024 * 1024, '\0'); 
  auto input_view = absl::string_view(input);
  auto actual = MakeCord(std::move(input));
  std::vector<absl::string_view> chunks{actual.chunk_begin(),
                                        actual.chunk_end()};
  EXPECT_THAT(
      chunks,
      ElementsAre(AllOf(
          Property("data", &absl::string_view::data, input_view.data()),
          Property("size", &absl::string_view::size, input_view.size()))));
```

Reply:
Yes, that what I thought you meant (modulo using `ResultOf` vs. `Property` for a type we do not own).  But such a test would depend on `std::string(std::move(input))` performing no copies.  And I was told that was not a good idea (it is not true for small strings, and there is not guarantee on the cutoff.  Granted, 4MiB seems large enough in any reasonable implementation).  

I can change my mind about this...

Reply:
Ack ack.  Which I think probably just leaves us with `MakeCord(std::string)`'s _"without copying"_ promise depending on `std::string(std::move(input))` performing no copies.  That is, either way we're testing/promising something we can't guarantee.  But I'll concede that it's not a practical concern.  Thanks.